### PR TITLE
Python builtin number conversion

### LINKED ISSUE #2524: Implement __int__ and __float__ for Variable
Implement these functions to convert scalar variables to int or float (https://docs.python.org/3/reference/datamodel.html#object.__int__).
Maybe also for single-element arrays (supported by numpy but conflicts with ours usage of `.value` vs `.values`)

Should it be `float(v) = float(v.value)` or `float(v) = v.astype(float).value`?

### PR DESCRIPTION
**Pull request by jl-wynen**
Fixes #2524

- I chose `float(v.value)` over `v.astype(float).value` because this way, we get the same conversion from `int` and `str` as for builtin types.
- Length-1 arrays are not supported in contrast with numpy because that would not match our usage of singular properties elsewhere (e.g. `.value`).

### CODE REVIEW

File: src/scipp/_binding.py [hunk 1]
```diff
@@ -30,6 +30,19 @@ def bind_get():
         setattr(cls, 'get', method)
 
 
+def _int_dunder(self) -> int:
+    return int(self.value)
```
Comment (SimonHeybrock) [inline src/scipp/_binding.py:34]: Shouldn't we also assert that self is dimensionless and has no variance?

Comment (jl-wynen) [reply src/scipp/_binding.py:34]: Good point! We should definitely check the variance. But I am not sure about the unit. `var.value` doesn't check either.

Comment (SimonHeybrock) [reply src/scipp/_binding.py:34]: Why would `var.value` check the unit? It returns the content of the `value` field? In contrast, what is added here converts a variable to a float, but `1.2 meter` is not a float.

**Comment by jl-wynen** [comment]
Looking at the CI errors. It seems like pybind11 triggers a conversion to `int` when slicing. Maybe this can be solved by changing the order of overloads?

File: lib/python/bind_slice_methods.h [hunk 2]
```diff
@@ -270,31 +270,10 @@ void bind_slice_methods(pybind11::class_<T, Ignored...> &c) {
...
-          expect_positional_index(index);
-          slicer<T>::template set<std::tuple<Dim, py::slice>>(
-              self, {self.dim(), index}, data);
-        });
-  c.def("__setitem__", &slicer<T>::template set<std::tuple<Dim, scipp::index>>);
-  c.def("__setitem__", &slicer<T>::template set<std::tuple<Dim, py::slice>>);
-  c.def("__setitem__", &slicer<T>::template set<py::ellipsis>);
+  // These overloads must also be placed before the ones accepting std::tuple.
+  // Otherwise, pybind11 would call `int()` on the 2nd tuple element
+  // because of `Variable.__int__`.
```
Comment (SimonHeybrock) [inline lib/python/bind_slice_methods.h:275]: What happens now if I do `da[sc.scalar(23)]`? Do we still get a dimension error, as we should?

File: pyproject.toml [hunk 3]
```diff
@@ -26,6 +26,8 @@ archs = ["x86_64", "arm64"]
 addopts = "-ra -v"
 filterwarnings = [
   'ignore:You are running a "Debug" build of scipp:',
+  'ignore:sc.matrix\(\) has been deprecated:DeprecationWarning',
+  'ignore:sc.matrices\(\) has been deprecated:DeprecationWarning',
```
Comment (SimonHeybrock) [inline pyproject.toml:30]: Can we fix these errors instead of ignoring them?

File: tests/variable_scalar_test.py [hunk 4]
```diff
@@ -28,3 +29,55 @@ def test_scalar_Variable_values_property_PyObject():
...
+def test_scalar_Variable_conversion_to_builtin_int_bad_unit():
+    var = sc.scalar(7, unit='m')
+    with pytest.raises(sc.UnitError):
+        int(var)
+
+
+def test_scalar_Variable_conversion_to_builtin_int_with_variance():
+    var = sc.scalar(7.0, variance=2.0)
+    with pytest.raises(sc.VariancesError):
+        int(var)
```
Comment (SimonHeybrock) [inline tests/variable_scalar_test.py:60]: Missing tests for `DimensionError` if the input is not a scalar?

File: lib/python/bind_slice_methods.h (hunk 2 above)
Comment (jl-wynen) [reply lib/python/bind_slice_methods.h:275]: No, this gives
```
scipp._scipp.core.DTypeError: Cannot extract elements based on condition with non-boolean dtype. If you intended to select a range based on a label you must specify the dimension.
```
But that is unchanged, the overload `(T &self, const Variable &condition)` takes precedence over label based slicing. Which, to be honest, I don't even see in this list at all.

Comment (SimonHeybrock) [reply lib/python/bind_slice_methods.h:275]: 👍

File: pyproject.toml (hunk 3 above)
Comment (jl-wynen) [reply pyproject.toml:30]: Not until we actually remove `matrix` and `matrices`. Unless you want to remove tests for them.

File: lib/python/bind_slice_methods.h (hunk 2 above)
Comment (jl-wynen) [reply lib/python/bind_slice_methods.h:275]: Given the complexity of `__getitem__` arguments, maybe we want to move to manually parsing them like in `Variable.__init__` in order to catch all edge cases and raise proper exceptions?

**Review by SimonHeybrock — APPROVED** [review]

File: lib/python/bind_slice_methods.h (hunk 2 above)
Comment (SimonHeybrock) [reply lib/python/bind_slice_methods.h:275]: 👍 
... but also never change a running system?

Comment (SimonHeybrock) [reply lib/python/bind_slice_methods.h:275]: But proper exception *would* be really useful.

**Pull request merged**